Gamemode Entities

DevulTj opened the pull request:
I've written a loose implementation for gamemode entities. This means that anything using IGamemodeEntity will register itself to a gamemode so the gamemode can access it.

They use target names (perhaps we should change this though) - so your entity's name should match the target name that you give your gamemode entity.

You can then either set up tags for your gamemode entities in Hammer: ```BaseGamemode.GetEntitiesWithTag( string tag )```

Or filter by type: ```BaseGamemode.GetEntitiesOfType( string tag = null )```

The idea is that we can set up entire gamemodes in Hammer. Ex: Set up camera positions using specific entities, and set tags for context.

DoctorGurke commented on code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:
```diff
@@ -0,0 +1,44 @@
+	}
+	public IEnumerable<T> GetEntitiesOfType<T>( string tag = null )
+	{
+		if ( tag is not null )
```
The logic here seems to be the wrong way around?

DevulTj replied:
Yeah, whoops

DoctorGurke commented on code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:
```diff
@@ -0,0 +1,44 @@
+			{
+				Entities.Add( entity );
+				gamemodeEntity.Gamemode = this;
+				entity.Parent = this;
```
good thinking with parenting them. We'll have to make sure map cleanup works properly along the line by parenting stuff to the game and/or using a custom cleanup filter. But that's just a random sidenote

DevulTj replied:
Yeah!

DoctorGurke commented on code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:
```diff
@@ -0,0 +1,44 @@
+	[Event.Entity.PostSpawn]
+	protected virtual void GatherTargetEntities()
+	{
+		var ents = FindAllByName( Name );
```
for the sake of I/O targetnames we should probably use a property on GamemodePointEntity that links them to the gamemode instead, similar to how we do it with the orbs.

DevulTj commented:
I've refactored some of it. It's now using an explicit property - and I've removed IGamemodeEntity as it was causing too many problems than I was trying to solve with it.

DoctorGurke commented on code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:
```diff
@@ -0,0 +1,43 @@
+	public IEnumerable<Entity> GetEntitiesWithTag( string tag )
+	{
+		return Entities.Where( x => x.Tags.Has( tag ) );
+	}
```
spacing here

DoctorGurke commented on code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:
```diff
@@ -0,0 +1,43 @@
+	{
+		var ents = All.OfType<GamemodePointEntity>();
+
+		Log.Info( $"[{Name}] Finding gamemode entities, count: {ents.Count()}" );
```
This should probably be Log.Debug

DoctorGurke commented on code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:
```diff
@@ -0,0 +1,43 @@
+		{
+			if ( gamemodeEntity.GamemodeId.ToLower() == Name.ToLower() )
+			{
+				Log.Info( $"	> Matched GamemodePointEntity for {Name}" );
```
This too

DevulTj replied:
This one is intended so it looks like this
![sbox_Tz2JySglv7](https://user-images.githubusercontent.com/22372616/169662611-a0092e2c-d772-4f07-b7b2-ad70f37ce351.png)

DevulTj replied:
Oh you meant log.debug mb
